### PR TITLE
Feature/metadata2

### LINKED ISSUE #4564: [question] How to annotate packages with visible free-form meta data
When packaging and distributing pre-built binaries for multiple configurations
I'd like to annotate the packages with free-form information the users or other tools
can query (conan search <reference>, conan get <reference> or conan inspect <reference>)
something like:

    description  # ok, this exists already
    topics: # ok, this exists already
    
    free_form_info: {"build_date": "10102019-1234",
                     "module_git_sha1": "1234567890",
                     "features": ["openssl", "zlib", ...]
                    }

In our use case the settings (customized) contain the os, arch, compiler etc.
mandatory settings. The other meta information related to the binaries are "good to know"
for the users so when users are installing the packages they will not do it like:

    conan install -s os=Linux -s arch=x86_64 -s feature_openssl=True -s feature_zlib=False

but like this:

    conan install -s os=Linux -s arch=x86_64

Is there any way to add custom meta information besides the "description" and "topics"?

### LINKED ISSUE #4689: Capture build log in package
I originally brought this up in https://github.com/conan-io/conan/issues/814 but now looking at the proposed solution https://github.com/conan-io/conan/pull/884 it is not what I expected. The idea was to automatically include the log in the package so that if you go to look at your package and something seems wrong (or you just want to confirm something), you can look at the log to get some insight. In other words, it gives you some traceability into how the package was created. Maybe the package build uses a tool (e.g. Cmake) and the log let's you know what version of cmake was used to build the package. Or a specific version of a compiler (where maybe the profile isn't as specific as you want to know).

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Feature: Metadata improvements.
Docs: Omit

- Creating metadata folders automatically, so easier for users
- Fixed bug in downloading for the first time with metadata
- Allowed re-downloading without "file already exists" error (required manual removal, very inconvenient)
- Make sure metadata is not cached in the ``download_cache`` (as it is not checksum, nor revision controlled, and it is mutable)
- Renamed interface ``pkg_metadata_folder`` => ``package_metadata_folder``
- Made available the metadata folders in ``ConanFileInterface`` for deployers access
- Initial POC of deployers access to metadata (missing recipe metadata)

Close https://github.com/conan-io/conan/issues/4689
Close https://github.com/conan-io/conan/issues/4564

### CODE REVIEW

File: conan/api/subapi/download.py [hunk 1]
```diff
@@ -62,5 +62,5 @@ def package(self, pref: PkgReference, remote: Remote, metadata=None):
         conan_file_path = layout.conanfile()
         conanfile = app.loader.load_basic(conan_file_path, display=pref.ref)
         output.info(f"Downloading package '{pref.repr_notime()}'")
-        app.remote_manager.get_package(conanfile, pref, remote)
+        app.remote_manager.get_package(conanfile, pref, remote, metadata)
```
Comment (memsharded) [inline conan/api/subapi/download.py]: this was a bug, forgetting to pass the argument resulted in metadata not downloaded.

File: conans/client/cmd/export.py [hunk 2]
```diff
@@ -39,7 +39,9 @@ def cmd_export(app, conanfile_path, name, version, user, channel, graph_lock=Non
     # TODO: cache2.0 move this creation to other place
     mkdir(export_folder)
     mkdir(export_src_folder)
-    conanfile.folders.set_base_recipe_metadata(recipe_layout.metadata())
+    recipe_metadata = recipe_layout.metadata()
+    mkdir(recipe_metadata)
```
Comment (memsharded) [inline conans/client/cmd/export.py:43]: Automatic creation of the folder, so ``conan cache path --folder=metadata`` has an existing folder to copy things to.

File: conans/client/downloaders/caching_file_downloader.py [hunk 3]
```diff
@@ -158,10 +158,10 @@ def __init__(self, requester, config, scope=None):
             raise ConanException("core.download:download_cache must be an absolute path")
         self._file_downloader = FileDownloader(requester, scope=scope)
 
-    def download(self, url, file_path, auth, verify_ssl, retry, retry_wait):
-        if not self._download_cache:
+    def download(self, url, file_path, auth, verify_ssl, retry, retry_wait, metadata=False):
+        if not self._download_cache or metadata:  # Metadata not cached and can be overwritten
```
Comment (memsharded) [inline conans/client/downloaders/caching_file_downloader.py:162]: The metadata cannot be cached, as it is not immutable, and not controlled by checksum

File: conans/client/installer.py [hunk 4]
```diff
@@ -326,12 +326,14 @@ def _handle_package(self, package, install_reference, remotes, handled_count, to
...
         # Make sure that all nodes with same pref compute package_info()
         pkg_folder = package_layout.package()
+        pkg_metadata = package_layout.metadata()
         assert os.path.isdir(pkg_folder), "Pkg '%s' folder must exist: %s" % (str(pref), pkg_folder)
         for n in package.nodes:
             n.prev = pref.revision  # Make sure the prev is assigned
             conanfile = n.conanfile
             # Call the info method
             conanfile.folders.set_base_package(pkg_folder)
+            conanfile.folders.set_base_pkg_metadata(pkg_metadata)
```
Comment (memsharded) [inline conans/client/installer.py:336]: This allows deployers to access ``dep.package_metadata_folder`` to collect and deploy metadata.

File: conans/client/rest/rest_client_v2.py [hunk 5]
```diff
@@ -39,20 +39,24 @@ def _get_file_list_json(self, url):
...
+
+        if not only_metadata:
+            accepted_files = ["conanfile.py", "conan_export.tgz", "conanmanifest.txt",
+                              "metadata/sign"]
+            files = [f for f in server_files if any(f.startswith(m) for m in accepted_files)]
+            # If we didn't indicated reference, server got the latest, use absolute now, it's safer
+            urls = {fn: self.router.recipe_file(ref, fn) for fn in files}
+            self._download_and_save_files(urls, dest_folder, files, parallel=True)
+            result.update({fn: os.path.join(dest_folder, fn) for fn in files})
+        if metadata:
```
Comment (memsharded) [inline conans/client/rest/rest_client_v2.py:53]: We need to separate the download to define ``metadata=True`` that allows 1) not caching and 2) allows overwriting existing files

File: conans/model/conan_file.py [hunk 6]
```diff
@@ -278,8 +278,8 @@ def recipe_metadata_folder(self):
         return self.folders.recipe_metadata_folder
 
     @property
-    def pkg_metadata_folder(self):
-        return self.folders.pkg_metadata_folder
+    def package_metadata_folder(self):
```
Comment (memsharded) [inline conans/model/conan_file.py:281]: Just a rename, to not use the ``pkg`` short version, everything else UI in Conan is ``package``.

**Review by AbrilRBS — APPROVED** [review]
Noting down the recipe metadata todos to have them for next release, but this looks good, thanks a lot for the explanation comments, made reviewing much easier :)

**Review by czoido — APPROVED** [review]

**Review by franramirez688 — APPROVED** [review]
👏

**Pull request merged**